multy_core/src/bitcoin/bitcoin_transaction_segwit.cpp

```diff
@@ -0,0 +1,506 @@
+                << " Address type: " << address_type;
+    }
+
+    if (expected_net_type != net_type)
+    {
+        THROW_EXCEPTION2(ERROR_INVALID_ADDRESS, "Wrong net type for address.")
+                << " Expected: " << expected_net_type
+                << " actual:" << net_type;
+    }
+
+    return make_script_pub_key(*binary_address, address_type);
+}
+
+
+
+} // namespace
+
+namespace multy_core
+{
+namespace internal
+{
+
+BitcoinTransactionSegWitDestination::BitcoinTransactionSegWitDestination(BitcoinNetType net_type)
+    : amount(m_properties,
+             "amount",
+             Property::REQUIRED,
+             &verify_bigger_than<BigInt, 0, ERROR_INVALID_ARGUMENT>),
+      address(m_properties,
+              "address",
+              Property::REQUIRED,
+              [this](const std::string &new_address) {
+                    this->sig_script = make_script_pub_key_from_address(this->m_net_type, new_address);
+              }),
+      is_change(m_properties,
+                "is_change",
+                Property::OPTIONAL,
+                [](const int32_t value) {
+                    if (value != 0 )
+                    {
+                        THROW_EXCEPTION("NOT SET is_change to segwit TX");
+                    }
+                }),
+      sig_script(),
+      m_net_type(net_type)
+{
+}
+
+//BinaryData BitcoinTransactionSegWitDestination::preview_output()
+//{
+//    BitcoinDataStream result;
+
+//    result << amount;
+//    result << sig_script;
+
+//    return result.get_content();
+//}
+
+BitcoinTransactionSegWitSource::BitcoinTransactionSegWitSource()
+    : prev_transaction_hash(
+              m_properties,
+              "prev_tx_hash",
+              Property::REQUIRED,
+              [](const BinaryData& new_tx_out_hash) {
+                  if (new_tx_out_hash.len != 32)
+                  {
+                      THROW_EXCEPTION2(
+                              ERROR_TRANSACTION_SOURCE_INVALID_PREV_TX_HASH,
+                              "Previous transaction hash should be"
+                              "exactly 32 bytes long.");
+                  }
+              }),
+      prev_transaction_out_index(
+              m_properties,
+              "prev_tx_out_index",
+              Property::REQUIRED),
+      prev_transaction_out_script_pubkey(
+              m_properties,
+              "prev_tx_out_script_pubkey",
+              Property::REQUIRED),
+      private_key(
+              m_properties,
+              "private_key",
+              Property::REQUIRED),
+      sequence(make_clone(BinaryData{BITCOIN_INPUT_SEQ_FINAL, 4}),
+                m_properties, "sequence", Property::OPTIONAL),
+      script_signature(),
+      script_witness(),
+      amount(m_properties, "amount", Property::REQUIRED)
+{
+}
+
+void BitcoinTransactionSegWitSource::sign(hash<256>& hash_prevouts, hash<256>& hash_sequence, hash<256>& hash_outputs, BinaryDataPtr& lock_time)
+{
+
```

Comment:
Well, it a big chunk of fancy code. Some reference to documentation (or reference implementation or both) would be nice.